### PR TITLE
Lecture 1 solutions

### PR DESCRIPTION
**Pull request by gecec**
### Solutions for Lecture 1

cc @chshersh @vrom911

### CODE REVIEW

File: src/Lecture1.hs [hunk 1]
```diff
@@ -63,7 +64,11 @@ sumOfSquares x y = error "TODO!"
 
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Int -> Int
+lastDigit inputNum = 
+    if (inputNum < 0) 
+    then mod ((-1)*inputNum) 10
+    else mod inputNum 10
```
Comment (chshersh) [inline src/Lecture1.hs]: That's correct! 💯 
You can simplify the implementation a bit using the standard function `abs` that returns an absolute value of a number:

```haskell
λ: abs 7
7
λ: abs (-7)
7
```

File: src/Lecture1.hs [hunk 2]
```diff
@@ -77,7 +82,21 @@ and 1 is the smallest, and 7 - 1 = 6.
 Try to use local variables (either let-in or where) to implement this
 function.
 -}
-minmax x y z = error "TODO"
+minmax :: Int -> Int -> Int -> Int
+minmax x y z = (max (max x y) z) - (min (min x y) z)
+    where 
+        max :: Int -> Int -> Int
+        max a b =
```
Comment (chshersh) [inline src/Lecture1.hs]: You don't need to implement `max` and `min` functions. These are already standard functions 👍🏻

File: src/Lecture1.hs [hunk 3]
```diff
@@ -94,7 +113,18 @@ start position can be considered as zero (e.g. substring from the
...
+    if ((end < 0) || (end < start))
+    then ""
+    else take (end - (evalStart start) + 1) (drop (evalStart start) str)
+
+    where 
+        evalStart :: Int -> Int
+        evalStart localStart = 
+            if (localStart < 0)
+            then 0
+            else localStart
```
Comment (chshersh) [inline src/Lecture1.hs]: You can write this function shorter by using standard function `max` 🙂 
```suggestion
        evalStart localStart = max 0 localStart
```

File: src/Lecture1.hs [hunk 4]
```diff
@@ -104,7 +134,16 @@ and finds a sum of the numbers inside this string.
...
-strSum str = error "TODO"
+strSum :: String -> Int
+strSum str = 
+    sum 0 (words str)
+    where
+        sum :: Int -> [String] -> Int
+        sum result numbers =
+            if null numbers
+                then result
+                else sum (result + (read (head numbers) :: Int)) (tail numbers)
```
Comment (chshersh) [inline src/Lecture1.hs]: `sum` is already a standard function in Haskell, you can use it instead.

Also, as you specify the top-level type signature for this function, the compiler would be able to figure out what type `read` function has in this case due to type inference, so you can skip specifying it in the body:

```suggestion
                else sum (result + (read (head numbers))) (tail numbers)
```

I see that the `result` variable is redundant in your implementation of `sum`. You don't change it so it's always `0` and doesn't affect the final outcome.

So if you use standard functions `map` and `sum`, you can implement `strSum` as a nice one-liner 🙂

File: src/Lecture1.hs [hunk 5]
```diff
@@ -119,4 +158,5 @@ and lower than 6 elements (4, 5, 6, 7, 8 and 9).
 
 🕯 HINT: Use recursion to implement this function.
 -}
-lowerAndGreater n list = error "TODO"
+lowerAndGreater :: Int->[Int]->String
+lowerAndGreater n list = ((show n) ++ " is greater than " ++ (show (length (filter (<n) list))) ++ " elements and lower than " ++ (show (length (filter (>n) list))) ++ " elements")
```
Comment (chshersh) [inline src/Lecture1.hs]: That's correct! 💯 

However, with this solution, you will walk through the list twice.
With the recursive solution and manual check on every element in the list on both conditions at the same time (without `filter` function), you would be able to write it more performant with only one traversal of the list 🙂

**Review by chshersh — APPROVED** [review]
Congrats! The first homework is complete 👏🏼 
Hope you enjoyed the course so far. Keep up the great work, and feel free to ask any questions if you have 👌🏼

File: src/Lecture1.hs (hunk 1 above)
Comment (gecec) [reply src/Lecture1.hs]: Replace with standard `abs`. Thx!

File: src/Lecture1.hs (hunk 2 above)
Comment (gecec) [reply src/Lecture1.hs]: Replaced

File: src/Lecture1.hs (hunk 4 above)
Comment (gecec) [reply src/Lecture1.hs]: Done!

File: src/Lecture1.hs (hunk 5 above)
Comment (gecec) [reply src/Lecture1.hs]: Replaced it with recursive function + auxiliary function to count - not very elegant :(

**Pull request merged**